pravega/pravega
# Issue 392: (SegmentStore) Evicting unused segment metadata.

**Change log description**
Segment Metadata eviction:
- **ContainerMetadata:** Added the ability to purge (evict) unused Segment Metadatas. This also involves keeping track of some segment-related usage statistics.
- **SegmentContainer:** 
    - Added a periodic CleanupTask to StreamSegmentContainer, which invokes the metadata cleanup, then cleans up the ReadIndex.
    - Removed **MetadataRepository** and InMemoryMetadataRepository. SegmentContainers create their own metadata objects and are responsible for their lifecycle.
- **StorageWriter:** Auto-cleaning up SegmentAggregators for those segments which have been deleted or evicted from the metadata.
- **ReadIndex:**
    - Updated cleanup() to consider all Read Indices for removal if no metadata or segments were deleted (if nothing passed as an arg).
    - Hitting the Metadata with every read to indicate a Segment is in use.

Other changes:
- **DurableDataLog:** truncate() returns Void and no longer the number of truncated items (this was not used).
- Merged uses of **AutoStopwatch** and **Timer** into Timer (which now derives from AbstractTimer to allow for easier unit testing).
- Various small changes to fix reported concurrency issues.
- Removed 'CacheUpdater' and merged its useful code into MemoryLogUpdater (that class losts its purpose once the Cache became fully controlled by the ReadIndex). 
- Briefly refactored ReadIndex to be the one creating and closing its caches (previously it was the container, which doesn't make sense anymore because the ReadIndex is the owner now).

**Purpose of the change**
A bunch of segment-related objects would unnecessarily linger in memory after their segments have been deleted or haven't been used for a while. When running a SegmentStoreService for a long time, these could accumulate in large numbers, use memory and slow down the metadata snapshotting (when it happens). 

Fixes #392 

**What the code does**
1. Keep track of a few usage statistics for Segments (last requested, last known sequence number (of an operation that references it)).
    - LastRequested: this gets updated in the SegmentContainer when we translate from SegmentName to SegmentId. It indicates the Segment has been accessed recently. This covers the ReadIndex and all the steps until the operation reaches the DurableLog.
    - Last Known Sequence Number: this is used to make sure that no operation that references the Segment is active in the system (It is compared to the current LowWaterMark in the Durable Log (the truncation SeqNo)). This covers the OperationProcessor, DurableLog and StorageWriter.
2. Periodically evict SegmentMetadata objects from the ContainerMetadata for those eligible Segments.
    - The check & eviction is done under the same lock that is used to request name-to-id mapping, which ensures that we won't serve an ID that is about to be evicted.
3. After the above-described eviction, auto-evict the Read Indices and Segment Aggregators (from the writer) for the same Segments.
    - This is done outside of any blocking lock (ReadIndex) and during the normal cycle of the Writer (it has a cleanup stage at the end)

All of these can be rebuilt the next time they are needed (the same way they were instantiated the first time).

**How to verify it**
New and existing unit tests must pass. Run SelfTest with lots of segments and lots of short-lived transactions and verify that the metadatas for those transactions are being evicted (via logs) and that no verification errors are detected.
